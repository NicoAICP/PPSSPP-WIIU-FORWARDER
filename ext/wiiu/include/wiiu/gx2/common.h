#pragma once

#define GX2_SCAN_BUFFER_ALIGNMENT      0x1000
#define GX2_SHADER_ALIGNMENT           0x100
#define GX2_CONTEXT_STATE_ALIGNMENT    0x100
#define GX2_UNIFORM_BLOCK_ALIGNMENT    0x100
#define GX2_DISPLAY_LIST_ALIGNMENT     0x20
#define GX2_VERTEX_BUFFER_ALIGNMENT    0x40
#define GX2_INDEX_BUFFER_ALIGNMENT     0x20

#define GX2_ENABLE                     TRUE
#define GX2_DISABLE                    FALSE

#define GX2_TRUE                       TRUE
#define GX2_FALSE                      FALSE

#ifndef GX2_COMP_SEL
#define GX2_COMP_SEL(c0, c1, c2, c3) (((c0) << 24) | ((c1) << 16) | ((c2) << 8) | (c3))
#ifdef __cplusplus
#include <functional>
#endif
#include <stdio.h>

#define _x 0
#define _y 1
#define _z 2
#define _w 3
#define _r 0
#define _g 1
#define _b 2
#define _a 3
#define _0 4
#define _1 5
#endif


#define GX2_DISABLE_WRAPS
