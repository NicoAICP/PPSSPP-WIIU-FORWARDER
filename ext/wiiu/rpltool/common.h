#ifndef COMMON_H
#define COMMON_H

#include <ctype.h>
#include <stdint.h>
#include <stdio.h>

typedef int8_t s8;
typedef int16_t s16;
typedef int32_t s32;

typedef uint8_t u8;
typedef uint16_t u16;
typedef uint32_t u32;

#define GET_ENUM_NAME(prefix, name, val) \
   case prefix##name: \
      return #name;

#define EMIT_ENUM_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(name type) { \
      switch (type) { cases } \
      return "unknown"; \
   }

#define GET_FLAG_NAME(prefix, name, val) \
   case prefix##name: \
      pos += snprintf(buffer + pos, sizeof(buffer) - pos, "%s%s", #name, seperator); \
      flags &= ~(1u << i); \
      break;

#define EMIT_FLAGS_TO_STR_FUNC(name, cases) \
   static inline const char *name##_to_str(int flags, const char *seperator) { \
      int i; \
      static char buffer[32 * 32]; \
      int pos = 0; \
      for (i = 0; i < 32; i++) \
         if ((1u << i) & flags) \
            switch (1u << i) { cases } \
      if (flags) \
         snprintf(buffer + pos, sizeof(buffer) - pos, "0x%08X", flags); \
      else if (pos) \
         buffer[pos - strlen(seperator)] = '\0'; \
      else \
         return "NULL"; \
      return buffer; \
   }

#define MAKE_MAGIC(a, b, c, d) (((a) << 24) | ((b) << 16) | ((c) << 8) | ((d) << 0))

#define log_var(X) printf("%-35s: 0x%0*X (%u)\n", #X, sizeof(X) * 2, (u32)(X), (u32)(X))
#define log_svar(X) printf("%-35s: 0x%0*X (%i)\n", #X, sizeof(X) * 2, (u32)(X), (s32)(X))
#define log_array(X) debug_array(X, sizeof(X), #X)
#define log_enum(type, X) printf("%-35s: 0x%X (%s)\n", #X, (X), type##_to_str(X))
#define log_flags(type, X) printf("%-35s: 0x%08X (%s)\n", #X, (X), type##_to_str(X, " | "))
#define log_magic(X) \
   printf("%-35s: '%c''%c''%c''%c' (0x%08X)\n", #X, (u32)(X) >> 24, (u32)(X) >> 16, (u32)(X) >> 8, (u32)(X), (u32)(X))
#define log_carray(X) debug_char_array(X, sizeof(X), #X)
#define log_str(X) printf("%-35s: \"%s\"\n", #X, (char *)(X))
#define log_off_str(X, data) printf("%-35s: 0x%08X \"%s\"\n", #X, (X), (char *)data + (X))
#define log_off_strs(X, data) \
   do { \
      printf("%-35s: 0x%08X ", #X, (X)); \
      const char *ptr = (char *)data + (X); \
      while (*ptr) { \
         ptr += printf("\"%s\" ", ptr) - 2; \
      } \
      printf("\n"); \
   } while (0)

static inline void dump_array(uint32_t *src, int size);
static inline void debug_array(uint32_t *src, int size, const char *name) {
   printf("%-35s:", name);
   dump_array(src, size);
}
static inline void debug_char_array(uint8_t *src, int size, const char *name) {
   int i;
   printf("%-35s: ", name);
   //   for (i = 0; i < size; i++)
   //      printf("%-2c ", isalnum(src[i])?src[i]: '?');
   //   printf("\n%-35s: ", "");
   for (i = 0; i < size; i++)
      printf("%02X ", src[i]);
   printf("\n");
}

static inline void dump_array(uint32_t *src, int size) {
   int i;
   size = (size / 4) - 1;
   printf("\n{");
   for (i = 0; i < size; i++) {
      if (!(i & 3))
         printf("\n   ");

      printf("0x%08X, ", src[i]);
   }

   if (!(i & 3))
      printf("\n   ");

   printf("0x%08X\n}\n", src[i]);
}

static inline u32 align_up(u32 x, u32 align) { return (x + (align - 1)) & ~(align - 1); }
static inline u32 align_down(u32 x, u32 align) { return x & ~(align - 1); }

#endif // COMMON_H
