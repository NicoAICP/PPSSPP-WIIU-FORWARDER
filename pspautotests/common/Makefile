TARGET_LIB = libcommon.a
EXTRA_TARGETS = libcommon_kernel.a
OBJS = common.o vram.o sysmem-imports.o
KERNEL_OBJS = common_kernel.o vram.o sysmem-imports.o

INCDIR = 
CFLAGS = -O0 -G 0 -g -Wall
CXXFLAGS = $(CFLAGS) -fno-exceptions -fno-rtti
ASFLAGS = $(CFLAGS)

LIBDIR =
LDFLAGS = -G0

PSPSDK=$(shell psp-config --pspsdk-path)
include $(PSPSDK)/lib/build.mak

%.o: %.S
	$(AS) $(ASFLAGS) -c -o $@ $<

common_kernel.o: common.c
	$(CC) $(CFLAGS) -DCOMMON_KERNEL=1 -c -o $@ $<

libcommon_kernel.a: $(KERNEL_OBJS)
	$(AR) cru $@ $(KERNEL_OBJS)
	$(RANLIB) $@
