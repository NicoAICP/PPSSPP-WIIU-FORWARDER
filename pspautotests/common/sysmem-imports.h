extern int sceKernelSetCompiledSdkVersion(int ver);
extern int sceKernelSetCompiledSdkVersion370(int ver);
extern int sceKernelSetCompiledSdkVersion380_390(int ver);
extern int sceKernelSetCompiledSdkVersion395(int ver);
extern int sceKernelSetCompiledSdkVersion401_402(int ver);
extern int sceKernelSetCompiledSdkVersion500_505(int ver);
extern int sceKernelSetCompiledSdkVersion507(int ver);
extern int sceKernelSetCompiledSdkVersion600_602(int ver);
extern int sceKernelSetCompiledSdkVersion603_605(int ver);
extern int sceKernelSetCompiledSdkVersion606(int ver);
extern int sceKernelGetCompiledSdkVersion();