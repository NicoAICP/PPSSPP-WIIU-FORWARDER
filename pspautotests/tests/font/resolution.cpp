#include <common.h>
#include <pspmodulemgr.h>
#include <pspiofilemgr.h>
#include <malloc.h>
extern "C" {
#include "libfont.h"
}

static FontLibraryHandle fontLib;
static FontLibraryHandle fontLibDone;
static int g_allocated = 0;
static int g_fontFile = 0;

static void *allocQuiet(void *data, u32 size) {
	g_allocated += size;
	void *p = malloc(size);
	//checkpoint("%08x = alloc(0x%x)", p, size);
	return p;
}

static void freeQuiet(void *data, void *p) {
	//checkpoint("free(%08x)", p);
	free(p);
}

bool loadFontModule() {
	checkpointNext("Init");
	if (RUNNING_ON_EMULATOR) {
		return true;
	}
	SceUID fontModule = sceKernelLoadModule("libfont.prx", 0, NULL);
	if (fontModule <= 0) {
		printf("TEST ERROR: Unable to load libfont.prx\n");
		return false;
	}

	int status = -1;
	int result = sceKernelStartModule(fontModule, 0, NULL, &status, NULL);
	if (result != fontModule || status != 0) {
		printf("TEST ERROR: libfont.prx startup failed (%08x, %08x)\n", result, status);
		return false;
	}
	return true;
}

void testSetResolution(const char *title, FontLibrary *lib, float h, float v) {
	int result = sceFontSetResolution(lib, h, v);
	if (result != 0) {
		checkpoint("%s: Failed (%08x -> %f, %f)", title, result, lib ? lib->hRes : NAN, lib ? lib->vRes : NAN);
	} else {
		checkpoint("%s: OK (%f, %f)", title, lib ? lib->hRes : NAN, lib ? lib->vRes : NAN);
	}
}

void testBasicParams() {
	checkpoint("Initial value: %04x", fontLib->altCharCode);

	checkpointNext("Libraries:");
	testSetResolution("  Normal", fontLib, 128.0f, 128.0f);
	testSetResolution("  Done", fontLibDone, 128.0f, 128.0f);
	testSetResolution("  NULL", NULL, 128.0f, 128.0f);

	checkpointNext("Values:");
	testSetResolution("  -1", fontLib, -1.0f, -1.0f);
	testSetResolution("  0", fontLib, 0.0f, 0.0f);
	testSetResolution("  1", fontLib, 1.0f, 1.0f);
	testSetResolution("  -1/128", fontLib, -1.0f, 128.0f);
	testSetResolution("  -INFINITY", fontLib, -INFINITY, -INFINITY);
	testSetResolution("  INFINITY", fontLib, INFINITY, INFINITY);
	// Crashes?
	//testSetResolution("  NAN", fontLib, NAN, NAN);
}

void testConversion(const char *title, float (*func)(FontLibraryHandle lib, float val, uint *errorCode)) {
	char temp[256];
	snprintf(temp, sizeof(temp), "%s conversions:", title);
	checkpointNext(temp);

	uint error;
	float result;

	result = func(fontLibDone, 5.0f, &error);
	checkpoint("  Lib done: %08x, %f", error, result);

	result = func(NULL, 5.0f, &error);
	checkpoint("  Lib NULL: %08x, %f", error, result);

	// Crashes.
	//result = func(fontLib, 5.0f, NULL);
	//checkpoint("  Missing error: %08x, %f", error, result);

	sceFontSetResolution(fontLib, 128.0f, 128.0f);
	result = func(fontLib, 5.0f, &error);
	checkpoint("  5.0f x 128.0f: %08x, %f", error, result);

	sceFontSetResolution(fontLib, 64.0f, 64.0f);
	result = func(fontLib, 5.0f, &error);
	checkpoint("  5.0f x 64.0f: %08x, %f", error, result);

	sceFontSetResolution(fontLib, 64.0f, 64.0f);
	result = func(fontLib, -5.0f, &error);
	checkpoint("  -5.0f x 64.0f: %08x, %f", error, result);

	sceFontSetResolution(fontLib, 64.0f, 64.0f);
	result = func(fontLib, 0.0f, &error);
	checkpoint("  0.0f x 64.0f: %08x, %f", error, result);
}

void testConversions() {
	testConversion("sceFontPixelToPointH", &sceFontPixelToPointH);
	testConversion("sceFontPixelToPointV", &sceFontPixelToPointV);
	testConversion("sceFontPointToPixelH", &sceFontPointToPixelH);
	testConversion("sceFontPointToPixelV", &sceFontPointToPixelV);
}

extern "C" int main(int argc, char *argv[]) {
	if (!loadFontModule()) {
		return 1;
	}

	FontNewLibParams libParams;
	memset(&libParams, 0, sizeof(libParams));
	libParams.allocFuncAddr = allocQuiet;
	libParams.freeFuncAddr = freeQuiet;
	libParams.numFonts = 4;

	uint error = -1;
	fontLib = sceFontNewLib(&libParams, &error);
	fontLibDone = sceFontNewLib(&libParams, &error);
	sceFontDoneLib(fontLibDone);

	testBasicParams();
	testConversions();

	sceFontDoneLib(fontLib);

	return 0;
}
