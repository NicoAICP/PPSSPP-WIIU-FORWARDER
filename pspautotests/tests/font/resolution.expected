[x] Init
[r] Initial value: 005f

[x] Libraries:
[x]   Normal: OK (128.000000, 128.000000)
[x]   Done: OK (128.000000, 128.000000)
[x]   NULL: Failed (80460002 -> nan, nan)

[x] Values:
[x]   -1: Failed (80460003 -> 128.000000, 128.000000)
[x]   0: Failed (80460003 -> 128.000000, 128.000000)
[x]   1: OK (1.000000, 1.000000)
[x]   -1/128: Failed (80460003 -> 1.000000, 1.000000)
[x]   -INFINITY: Failed (80460003 -> 1.000000, 1.000000)
[x]   INFINITY: OK (inf, inf)

[x] sceFontPixelToPointH conversions:
[x]   Lib done: 00000000, 2.812500
[x]   Lib NULL: 80460002, 0.000000
[x]   5.0f x 128.0f: 00000000, 2.812500
[x]   5.0f x 64.0f: 00000000, 5.625000
[x]   -5.0f x 64.0f: 00000000, -5.625000
[x]   0.0f x 64.0f: 00000000, 0.000000

[x] sceFontPixelToPointV conversions:
[x]   Lib done: 00000000, 2.812500
[x]   Lib NULL: 80460002, 0.000000
[x]   5.0f x 128.0f: 00000000, 2.812500
[x]   5.0f x 64.0f: 00000000, 5.625000
[x]   -5.0f x 64.0f: 00000000, -5.625000
[x]   0.0f x 64.0f: 00000000, 0.000000

[x] sceFontPointToPixelH conversions:
[x]   Lib done: 00000000, 8.888889
[x]   Lib NULL: 80460002, 0.000000
[x]   5.0f x 128.0f: 00000000, 8.888889
[x]   5.0f x 64.0f: 00000000, 4.444445
[x]   -5.0f x 64.0f: 00000000, -4.444445
[x]   0.0f x 64.0f: 00000000, 0.000000

[x] sceFontPointToPixelV conversions:
[x]   Lib done: 00000000, 8.888889
[x]   Lib NULL: 80460002, 0.000000
[x]   5.0f x 128.0f: 00000000, 8.888889
[x]   5.0f x 64.0f: 00000000, 4.444445
[x]   -5.0f x 64.0f: 00000000, -4.444445
[x]   0.0f x 64.0f: 00000000, 0.000000
